// bricklink-api/dist/catalogItem/catalogItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CatalogItem = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Represents a Catalog Item.
 */
var CatalogItem = exports.CatalogItem = function () {
  /**
   * Create a new instance of a Catalog Item
   * @param {object} [data] Data typically returned from an API response.
   */
  function CatalogItem(data) {
    _classCallCheck(this, CatalogItem);

    data = data || {};

    /** @type {string} */
    this.no = data.no || '';
    /** @type {string} */
    this.name = data.name || '';
    /** @type {string} */
    this.type = data.type || '';
    /** @type {number} */
    this.category_id = data.category_id || data.categoryID || 0;
    /** @type {string} */
    this.alternate_no = data.alternate_no || '';
    /** @type {string} */
    this.image_url = data.image_url || '';
    /** @type {string} */
    this.thumbnail_url = data.thumbnail_url || '';
    /** @type {number} */
    this.weight = data.weight || '';
    /** @type {string} */
    this.dim_x = data.dim_x || '';
    /** @type {string} */
    this.dim_y = data.dim_y || '';
    /** @type {string} */
    this.dim_z = data.dim_z || '';
    /** @type {string} */
    this.year_released = data.year_released || '';
    /** @type {string} */
    this.description = data.description || '';
    /** @type {boolean} */
    this.is_obsolete = data.is_obsolete || false;
    /** @type {string} */
    this.language_code = data.language_code || '';
  }

  /**
   * Method to get a known catalog item
   *
   * Usage:
   *
   * ```
   * var req = CatalogItem.get(ItemType.Part, '3001');
   * client.send(req).then(item => console.log(item));
   * ```
   *
   * @param {string} itemType Catalog item type i.e. - MINIFIG, PART, SET, BOOK, GEAR, CATALOG, INSTRUCTION, UNSORTED_LOT, ORIGINAL_BOX
   * @param {string} itemNumber Catalog item number
   * @return {Request} A request that is ready to execute with a client.
   */


  _createClass(CatalogItem, null, [{
    key: 'get',
    value: function get(itemType, itemNumber) {
      var method = _request.Request.GET;
      var uri = '/items/' + itemType + '/' + itemNumber;

      return new _request.Request(method, uri, null, function (data) {
        var item = new CatalogItem(data);
        return item;
      });
    }
  }]);

  return CatalogItem;
}();
