// bricklink-api/dist/catalogItem/priceGuide.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PriceGuideOptions = exports.PriceGuide = exports.PriceDetail = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

var _catalogItem = require('./catalogItem');

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Specific sale price details for a givent item.
 */
var PriceDetail =
/**
 * Create a Price Detail instance.
 * @param {object} [data] Data returned from an API response.
 */
exports.PriceDetail = function PriceDetail(data) {
  _classCallCheck(this, PriceDetail);

  data = data || {};
  /** @type {number} */
  this.quantity = data.quantity || 0;
  /** @type {number} */
  this.unit_price = data.unit_price || 0;
  /** @type {string} */
  this.shipping_available = data.shipping_available || '';
  /** @type {string} */
  this.shipping_available = data.shipping_available || '';
  /** @type {string} */
  this.seller_country_code = data.seller_country_code || '';
  /** @type {string} */
  this.buyer_country_code = data.buyer_country_code || '';
  /** @type {Date|null} */
  this.date_ordered = data.date_ordered ? new Date(datea.date_ordered) : null;
};

/**
 * Represents a Price Guide for a catalog item.
 */


var PriceGuide = exports.PriceGuide = function () {
  /**
   * Create a new instance of a price guide.
   * @param {object} [data] Data returned from an API response.
   */
  function PriceGuide(data) {
    _classCallCheck(this, PriceGuide);

    /** @type {Item} item The item that belongs to the price guide. */
    this.item = data.item ? new _catalogItem.CatalogItem(data.item) : new Item();
    /** @type {string} new_or_used Whether or not the price guide is new or used condition. */
    this.new_or_used = data.new_or_used || '';
    /** @type {string} */
    this.currency_code = data.currency_code || '';
    /** @type {number} */
    this.min_price = data.min_price || 0;
    /** @type {number} */
    this.max_price = data.max_price || 0;
    /** @type {number} */
    this.avg_price = data.avg_price || 0;
    /** @type {number} */
    this.qty_avg_price = data.qty_avg_price || 0;
    /** @type {number} */
    this.unit_quantity = data.unit_quantity || 0;
    /** @type {number} */
    this.total_quantity = data.total_quantity || 0;
    /** @type {PriceDetail[]} */
    this.price_detail = data.price_detail ? data.price_detail.map(function (detail) {
      return new PriceDetail(detail);
    }) : [];
  }

  /**
   * Method to get a known catalog item's price guide
   *
   * Usage:
   *
   * ```
   * var req = PriceGuide.get(ItemType.Part, '3001', {new_or_used: Condition.Used});
   * client.send(req).then(guide => console.log(guide));
   * ```
   *
   * @param {string} itemType Catalog item type as found at {@link ItemType}.
   * @param {string} itemNumber Catalog item number
   * @param {object} [params] Params as outlined in {@link PriceGuideOptions}.
   * @return {Request} A request that is ready to execute with a client.
   */


  _createClass(PriceGuide, null, [{
    key: 'get',
    value: function get(itemType, itemNumber, params) {
      params = params ? new PriceGuideOptions(params) : new PriceGuideOptions();
      var method = _request.Request.GET;
      var uri = 'items/' + itemType + '/' + itemNumber + '/price';

      return new _request.Request(method, uri, params, function (data) {
        return new _catalogItem.CatalogItem(data);
      });
    }
  }]);

  return PriceGuide;
}();

/**
 * Price guide option request params.
 */


var PriceGuideOptions = exports.PriceGuideOptions = function (_RequestParams) {
  _inherits(PriceGuideOptions, _RequestParams);

  /**
   * Create an instance of Price Guide options.
   * @param {object} data Optional parameter data.
   * @param {number|null} [data.color_id=null] The color Identification numbe of the item
   * @param {string} [data.guide_type='stock'] Indicates that which statistics to be provided. Options include "sold" and "stock"
   * @param {string} [data.new_or_used='N'] Indicates the condition of items that are included in the statistics. Acceptable values are: "N": new item (default) and "U": used item
   * @param {string|null} [data.country_code=null]
   * @param {string|null} [data.region=null]
   * @param {string|null} [data.currency_code=null]
   * @param {string} [data.vat='N']
   */
  function PriceGuideOptions(data) {
    _classCallCheck(this, PriceGuideOptions);

    var _this = _possibleConstructorReturn(this, (PriceGuideOptions.__proto__ || Object.getPrototypeOf(PriceGuideOptions)).call(this, data));

    data = data || {};
    /** @type {number|null} [data.color_id=null] The color Identification numbe of the item */
    _this.color_id = data.color_id || null;
    /** @type {string} [guide_type='stock'] Indicates that which statistics to be provided. Options include "sold" and "stock" */
    _this.guide_type = data.guide_type || 'stock';
    /** @type {string} [new_or_used='N'] Indicates the condition of items that are included in the statistics. Acceptable values are: "N": new item (default) and "U": used item */
    _this.new_or_used = data.new_or_used || 'N';
    /** @type {string|null} [country_code=null] */
    _this.country_code = data.country_code || null;
    /** @type {string|null} [region=null] */
    _this.region = data.region || null;
    /** @type {string|null} [currency_code=null] */
    _this.currency_code = data.currency_code || null;
    /** @type {string} [vat='N'] */
    _this.vat = data.vat || 'N';
    return _this;
  }

  return PriceGuideOptions;
}(_request.RequestParams);
