// bricklink-api/dist/catalogItem/subsets.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SubsetOptions = exports.Subset = exports.SubsetEntry = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

var _catalogItem = require('./catalogItem');

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Represents a subset entry.
 */
var SubsetEntry =
/**
 * Create an instance of a subset entry
 * @param {object} [data] The API response data
 * @param {object} [data.item] Item information
 * @param {number} [data.color_id] The color id of the entry item.
 * @param {number} [data.quantity] The number of items in the subset entry.
 * @param {number} [data.extra_quantity] The number of extra items included in the subset.
 * @param {boolean} [data.is_alternate] Indicator that the item is an alternate.
 */
exports.SubsetEntry = function SubsetEntry(data) {
  _classCallCheck(this, SubsetEntry);

  data = data || {};
  /** @type {CatalogItem} */
  this.item = data.item ? new _catalogItem.CatalogItem(data.item) : new _catalogItem.CatalogItem();
  /** @type {number} */
  this.color_id = data.color_id || 0;
  /** @type {number} */
  this.quantity = data.quantity || 0;
  /** @type {number} */
  this.extra_quantity = data.extra_quantity || 0;
  /** @type {boolean} */
  this.is_alternate = data.is_alternate || false;
};
/**
 * Represents a subset item. A subset item can have more than one matching entry due to alternate parts.
 */


var Subset = exports.Subset = function () {
  /**
   * Create an instance of a subset item.
   * @param {object} [data] The response data from API request.
   * @param {number} [data.match_no] The number of matching entries or 0 if there is no matching of alternative item.
   * @param {object} [data.entries] The specific entries for the subset item.
   */
  function Subset(data) {
    _classCallCheck(this, Subset);

    data = data || {};
    /** @type {number} */
    this.match_no = data.match_no || 0;

    var entries = data.entries || [];
    /** @type {SubsetEntry[]} */
    this.entries = entries.map(function (e) {
      return new SubsetEntry(e);
    });
  }
  /**
   * Method to get a subset of a catalog item.
   *
   * Usage:
   *
   * ```
   * var req = Subset.get(ItemType.Set, '6020-1', {break_minifigs: true});
   * client.send(req).then(subset => console.log(subset));
   * ```
   *
   * @param {string} itemType Catalog item type
   * @param {string} itemNumber Catalog item number
   * @param {object} options Options that conform to {@link SubsetOptions}.
   */


  _createClass(Subset, null, [{
    key: 'get',
    value: function get(itemType, itemNumber, options) {
      var method = _request.Request.GET;
      options = options ? new SubsetOptions(options) : new SubsetOptions();

      var uri = '/items/' + itemType + '/' + itemNumber + '/subsets';

      return new _request.Request(method, uri, options, function (data) {
        return data.map(function (e) {
          return new Subset(e);
        });
      });
    }
  }]);

  return Subset;
}();

/**
 * Options that can be used when make a request for a subset.
 */


var SubsetOptions = exports.SubsetOptions = function (_RequestParams) {
  _inherits(SubsetOptions, _RequestParams);

  /**
   * Make an instance of SubsetOptions.
   * @param {object} [data] Raw options object.
   * @param {number} [data.color_id] The color of the item(This value is valid only for {@link ItemType}.Part.
   * @param {boolean} [data.box] Indicates whether the set includes the original box
   * @param {boolean} [data.instruction] Indicates whether the set includes the original instruction
   * @param {boolean} [data.break_minifigs]	Indicates whether the result breaks down minifigs as parts
   * @param {boolean} [data.break_subsets] Indicates whether the result breaks down sets in set
   */
  function SubsetOptions(data) {
    _classCallCheck(this, SubsetOptions);

    var _this = _possibleConstructorReturn(this, (SubsetOptions.__proto__ || Object.getPrototypeOf(SubsetOptions)).call(this, data));

    data = data || {};
    /** @type {number|null} */
    _this.color_id = data.color_id || null;
    /** @type {boolean|null} */
    _this.box = data.box || null;
    /** @type {boolean|null} */
    _this.instruction = data.instruction || null;
    /** @type {boolean|null} */
    _this.break_minifigs = data.break_minifigs || null;
    /** @type {boolean|null} */
    _this.break_subsets = data.break_subsets || null;
    return _this;
  }

  return SubsetOptions;
}(_request.RequestParams);
