// bricklink-api/dist/catalogItem/supersets.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SupersetOptions = exports.Superset = exports.SupersetEntry = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

var _catalogItem = require('./catalogItem');

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Represents entries of a superset item.
 */
var SupersetEntry =
/**
 * Create an instance of a subset entry
 * @param {object} [data] The API response data
 * @param {object} [data.item] Item information
 * @param {number} [data.quantity] The number of items in the subset entry.
 * @param {string} [data.appear_as] How the entry appears in the superset item.
 */
exports.SupersetEntry = function SupersetEntry(data) {
  _classCallCheck(this, SupersetEntry);

  data = data || {};
  /** @type {CatalogItem} */
  this.item = data.item ? new _catalogItem.CatalogItem(data.item) : new _catalogItem.CatalogItem();
  /** @type {number} */
  this.quantity = data.quantity || 0;
  /** @type {string} */
  this.appear_as = data.appear_as || '';
};

/**
 * Represents a super set entry.
 */


var Superset = exports.Superset = function () {
  /**
   * Create an instance of a super set item.
   * @param {object} [data] The API response data
   * @param {object[]} [data.entries] Entries for a superset item.
   * @param {number} [data.color_id] The color id of the superset item.
   */
  function Superset(data) {
    _classCallCheck(this, Superset);

    data = data || {};
    /** @type {number} */
    this.color_id = data.color_id || 0;

    var entries = data.entries || [];
    /** @type {SupersetEntry[]} */
    this.entries = entries.map(function (e) {
      return new SupersetEntry(e);
    });
  }

  /**
   * Method to get a superset of a catalog item
   *
   * Usage:
   *
   * ```
   * var req = Superset.get(ItemType.Part, '4593');
   * client.send(req).then(superset => console.log(superset));
   * ```
   *
   * @param {string} itemType Catalog item type
   * @param {string} itemNumber Catalog item number
   * @param {object} options Options that conform to {@link SupersetOptions}.
   * @return {Request} A request that will return an array of {@link Superset} items.
   */


  _createClass(Superset, null, [{
    key: 'get',
    value: function get(itemType, itemNumber, options) {
      options = options ? new SupersetOptions(options) : new SupersetOptions();
      var method = _request.Request.GET;
      var uri = '/items/' + itemType + '/' + itemNumber + '/supersets';

      return new _request.Request(method, uri, options, function (data) {
        return data.map(function (set) {
          return new Superset(set);
        });
      });
    }
  }]);

  return Superset;
}();

/**
 * Options that can be used when make a request for a subset.
 */


var SupersetOptions = exports.SupersetOptions = function (_RequestParams) {
  _inherits(SupersetOptions, _RequestParams);

  /**
   * Make an instance of SubsetOptions.
   * @param {object} [data] Raw options object.
   * @param {number} [data.color_id] The color of the item.
   */
  function SupersetOptions(data) {
    _classCallCheck(this, SupersetOptions);

    var _this = _possibleConstructorReturn(this, (SupersetOptions.__proto__ || Object.getPrototypeOf(SupersetOptions)).call(this, data));

    data = data || {};
    /** @type {number|null} */
    _this.color_id = data.color_id || null;
    return _this;
  }

  return SupersetOptions;
}(_request.RequestParams);
