// bricklink-api/dist/oAuthHelper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OAuthHelper = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _crypto = require('crypto');

var _crypto2 = _interopRequireDefault(_crypto);

var _oauthSignature = require('oauth-signature');

var _oauthSignature2 = _interopRequireDefault(_oauthSignature);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * A helper for OAuth 1.0 header information and signing keys.
 */
var OAuthHelper = exports.OAuthHelper = function () {
  /**
   * Intialize the helper
   * @param {string} consumerKey The consumer key
   * @param {string} token The token value
   */
  function OAuthHelper(consumerKey, token) {
    _classCallCheck(this, OAuthHelper);

    /** @type {string} */
    this.consumerKey = consumerKey;
    /** @type {string} */
    this.token = token;
    /** @type {string} */
    this.nonce = _crypto2.default.randomBytes(16).toString('hex');
    /** @type {string} */
    this.signature = '';
    /** @type {number} */
    this.timestamp = Math.round(Date.now() / 1000).toString();
  }

  _createClass(OAuthHelper, [{
    key: 'sign',


    /**
     * Sign the request with oath headers and sets the signature member of the OauthHelper instance.
     */
    value: function sign(url, request, consumerSecret, tokenSecret) {
      var params = Object.assign({}, this.parameters, request.params.toObject());

      this.signature = _oauthSignature2.default.generate(request.method, url, params, consumerSecret, tokenSecret);
    }
  }, {
    key: 'parameters',
    get: function get() {
      return {
        oauth_consumer_key: this.consumerKey,
        oauth_token: this.token,
        oauth_nonce: this.nonce,
        oauth_timestamp: this.timestamp,
        oauth_signature_method: 'HMAC-SHA1',
        oauth_version: '1.0'
      };
    }
  }, {
    key: 'header',
    get: function get() {
      var str = 'OAuth ';

      var params = Object.assign({ 'realm': '' }, this.parameters);
      params['oauth_signature'] = this.signature;

      var keys = Object.keys(params);
      var pairs = keys.map(function (k) {
        return k + '="' + params[k] + '"';
      });

      str += pairs.join(',');
      return str;
    }
  }, {
    key: 'queryString',
    get: function get() {
      var params = Object.assign({}, this.parameters);
      params['oauth_signature'] = this.signature;
      var str = JSON.stringify(params);
      return "Authorization=" + str;
    }
  }]);

  return OAuthHelper;
}();
