// bricklink-api/dist/request.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Creates a request to use with the client.
 */
var Request = exports.Request = function () {
  /**
   * Create a new request.
   * @param {string} method The type of HTTP request to perform.
   * @param {string} uri The relative of full path uri of a request to perform. This should not include any query parameters.
   * @param {RequestParams} [params] Additional parameters to include with the request in either post body or query params.
   * @param {function} [callback=null] A callback function to perform after the request has been successfully performed.
   */
  function Request(method, uri, params, callback) {
    _classCallCheck(this, Request);

    /** @type {string} */
    this.method = method;
    /** @type {string} */
    this.uri = uri;
    /** @type {RequestParams} */
    this.params = params || new RequestParams();
    /** @type {function|null} */
    this.callback = callback || null;
  }
  /** @type {string} */


  _createClass(Request, null, [{
    key: 'GET',
    get: function get() {
      return 'get';
    }

    /** @type {string} */

  }, {
    key: 'PUT',
    get: function get() {
      return 'put';
    }

    /** @type {string} */

  }, {
    key: 'POST',
    get: function get() {
      return 'post';
    }

    /** @type {string} */

  }, {
    key: 'DELETE',
    get: function get() {
      return 'put';
    }
  }]);

  return Request;
}();

/**
 * Represents any query parameters to use with a request.
 */


var RequestParams = exports.RequestParams = function () {
  function RequestParams() {
    _classCallCheck(this, RequestParams);
  }

  _createClass(RequestParams, [{
    key: 'toQueryString',

    /**
     * Convert params for the request to a query string.
     * @return {string} parameters in the format of '?<param>=<value>'.
     */
    value: function toQueryString() {
      var pairs = [];
      for (var member in this) {
        if (this[member] != null) {
          pairs.push(member + '=' + this[member]);
        }
      }
      if (pairs.length > 0) {
        return '?' + pairs.join('&');
      }
      return '';
    }

    /**
     * Convert params for the request to a an object.
     * @return {object} trimmed down parameters object.
     */

  }, {
    key: 'toObject',
    value: function toObject() {
      var params = {};
      for (var member in this) {
        if (this[member] != null) {
          params[member] = this[member];
        }
      }
      return params;
    }
  }]);

  return RequestParams;
}();
