// bricklink-api/dist/color/color.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Color = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Represents a color object
 */
var Color = exports.Color = function () {
  /**
   * Create a new instance of a color
   * @param {object} [data] API response data.
   * @param {number} [data.color_id=0] The color id. i.e. - 10
   * @param {string} [data.color_name] The color name. i.e. - Dark Gray
   * @param {string} [data.color_code] The HTML HEX associated to the color. i.e. - 6b5a5a
   * @param {string} [data.color_type] A group that the color belongs to. i.e. - Solid, Transparent, etc...
   */
  function Color(data) {
    _classCallCheck(this, Color);

    data = data || {};
    /** @type {number} */
    this.color_id = data.color_id || 0;
    /** @type {string} */
    this.color_name = data.color_name || '';
    /** @type {string} */
    this.color_code = data.color_code || '';
    /** @type {string} */
    this.color_type = data.color_type || '';
  }

  /**
   * Method to get specific color details
   *
   * Usage:
   *
   * ```
   * var req = Color.get(10);
   * client.send(req).then(color => console.log(color));
   * ```
   *
   * @param {number} colorId a color id.
   * @return {Request} A request that resolves to a {@link Color} instance.
   */


  _createClass(Color, null, [{
    key: 'get',
    value: function get(colorId) {
      var method = _request.Request.GET;
      var uri = '/colors/' + colorId;

      return new _request.Request(method, uri, null, function (data) {
        return new Color(data);
      });
    }

    /**
     * Get a list of all colors
     *
     * Usage:
     *
     * ```
     * var req = Color.all();
     * client.send(req).then(colors => console.log(colors));
     * ```
     *
     * @return {Request} A request that resolves to an array of {@link Color}.
     */

  }, {
    key: 'all',
    value: function all() {
      var method = _request.Request.GET;
      var uri = '/colors';

      return new _request.Request(method, uri, null, function (data) {
        return data.map(function (d) {
          return new Color(d);
        });
      });
    }
  }]);

  return Color;
}();
